### ISSUE TITLE
ECDH not working

### ISSUE THREAD

**Issue by popovec** [OpenSC/OpenSC#1609]
ECDH operation (myeid driver) with opensc from git, commit 20daced60575c508c81b3fbb3ea2209b138b71f7.

pkcs11-tool  --login --pin XXXXXXXX --derive -m ECDH1-DERIVE --id 10   --input-file tmp/ecdh-bob-pub.der   --output-file tmp/ecdh-bob-alices-shared.der
error: PKCS11 function C_DeriveKey failed: rv = CKR_GENERAL_ERROR (0x5) Aborting.

APDUs from debug.. 
00 A4 08 0C 04 50 15 4B 01                   < select file
00 22 41 A4 0A 80 01 04 81 02 4B 01 84 01 00 < set sec envir.
00 2A 80 86 32 00 04 CC 99 26 38 32 9A ....  < operation decrypt! not general auth

Last working commit: be2cc38565c040af0a36b671d59877984678aa19 generates correct APDU:
00 86 00 00 35 7C 33 85 31 04 8C D7 .... < general authenticate

Wrong APDU is selected by this condition (src/libopensc/card-myeid.c):

        if (priv->sec_env && priv->sec_env->algorithm == SC_ALGORITHM_EC
                && priv->sec_env->operation == SC_SEC_OPERATION_DERIVE
                && priv->sec_env->algorithm_flags & SC_ALGORITHM_ECDSA_RAW)

This fail on priv->sec_env->algorithm_flags & SC_ALGORITHM_ECDSA_RAW, because priv->sec_env->algorithm_flags is zero. 

priv->sec_env->algorithm_flags comes from padding.c:511:sc_get_encoding_flags:

Relevant part from log:

Wrong: (commit e5707b545e5a2dc33b0ca52a8bf63f36f71b3d85)
padding.c:472:sc_get_encoding_flags: iFlags 0x0, card capabilities 0x80300300
padding.c:511:sc_get_encoding_flags: pad flags 0x0, secure algorithm flags 0x0
padding.c:512:sc_get_encoding_flags: returning with: 0 (Success)

Correct: (commit be2cc38565c040af0a36b671d59877984678aa19)
padding.c:288:sc_get_encoding_flags: iFlags 0x0, card capabilities 0x80300300
padding.c:322:sc_get_encoding_flags: pad flags 0x0, secure algorithm flags 0x100000
padding.c:323:sc_get_encoding_flags: returning with: 0 (Success)

sc_get_encoding_flags() is in src/libopensc/padding.c and this function was widely changed by commit e5707b545e5a2dc33b0ca52a8bf63f36f71b3d85

Here is point, where secure algorithm flags is set to 0:
```
sc_get_encoding_flags( sc_context_t *ctx,unsigned long iflags,...)

if ((caps & iflags) == iflags) {
                /* Card supports the signature operation we want to do, great, let's
                 * go with it then. */
                *sflags = iflags;
                *pflags = 0;

        } else ...
```
Here iflags is function parameter, in this case is set to 0, and this leads to *sflags = 0

I am not sure if iflags value is wrong, or function sc_get_encoding_flags() is wrong.  Old code sc_get_encoding_flags() sets *sflags to correct value by this part of code:
```
               /* Work with RSA, EC and maybe GOSTR? */
                if (!(caps & SC_ALGORITHM_RAW_MASK))
                        LOG_TEST_RET(ctx, SC_ERROR_NOT_SUPPORTED, "raw encryption is not supported");

                *sflags |= (caps & SC_ALGORITHM_RAW_MASK); /* adds in the one raw type */
                *pflags = 0;
```
But this part of code was removed in commit e5707b545e5a2dc33b0ca52a8bf63f36f71b3d85. 

One of the options to fix this issue is return back old code to function  sc_get_encoding_flags(). Second option is changing iflags value.

From where comes iflags = 0 ?

In framework-pkcs15.c, function pkcs15_prkey_derive() line 4207, flags is set to 0, then this value is used in sc_pkcs15_derive() call and this function then does call to sc_get_encoding_flags(). 

Below patch, but I am not sure, if this is really correct place to fix this problem:

```
diff --git a/src/pkcs11/framework-pkcs15.c b/src/pkcs11/framework-pkcs15.c
index c7c8f1b6..112f903e 100644
--- a/src/pkcs11/framework-pkcs15.c
+++ b/src/pkcs11/framework-pkcs15.c
@@ -4241,10 +4241,11 @@ pkcs15_prkey_derive(struct sc_pkcs11_session *session, void *obj,
                case SC_PKCS15_TYPE_PRKEY_EC:
                {
                        CK_ECDH1_DERIVE_PARAMS * ecdh_params = (CK_ECDH1_DERIVE_PARAMS *) pParameters;
                        ulSeedDataLen = ecdh_params->ulPublicDataLen;
                        pSeedData = ecdh_params->pPublicData;
+                       flags = SC_ALGORITHM_ECDSA_RAW;
                }
                break;
        }
 
        rv = sc_pkcs15_derive(fw_data->p15_card, prkey->prv_p15obj, flags,
```

**Comment by dengert** [comment]
There are two flags, SC_ALGORITHM_ECDSA_RAW and SC_ALGORITHM_ECDH_CDH_RAW. which are independent.  One for ECDSA and one for ECDH  It looks like the problem may be the new code is testing the wrong code in: 

```
  if (priv->sec_env && priv->sec_env->algorithm == SC_ALGORITHM_EC
            && priv->sec_env->operation == SC_SEC_OPERATION_DERIVE
            && priv->sec_env->algorithm_flags & SC_ALGORITHM_ECDSA_RAW)
```
Can you try changing the above to:
```
  if (priv->sec_env && priv->sec_env->algorithm == SC_ALGORITHM_EC
            && priv->sec_env->operation == SC_SEC_OPERATION_DERIVE
            && priv->sec_env->algorithm_flags & SC_ALGORITHM_ECDH_CDH_RAW)
```

**Comment by popovec** [comment]
This cannot work because priv->sec_env->algorithm_flags is still 0. Of course, flag SC_ALGORITHM_ECDH_CDH_RAW can be used in patch above, and then myeid driver must be changed as you suggest. 

When I was looking for SC_ALGORITHM_ECDH_CDH_RAW, I also found this:
src/libopensc/opensc.h
```
#define SC_ALGORITHM_RAW_MASK  (SC_ALGORITHM_RSA_RAW,SC_ALGORITHM_GOSTR3410_RAW,SC_ALGORITHM_ECDSA_RAW)
```

SC_ALGORITHM_ECDH_CDH_RAW is missing in this definition .. is it okay?

**Comment by dengert** [comment]
Maybe. But SC_ALGORITHM_RAW_MASK is not used anywhere. 

I just find it strange that card-myeid.c is testing for SC_ALGORITHM_ECDSA_RAW in the middle of deciding if it should call myeid_ecdh_derive. 

But card-sc-hsm.c does something similar at line 1008.  So I would suspect that the https://github.com/OpenSC/OpenSC/commit/e5707b545e5a2dc33b0ca52a8bf63f36f71b3d85 may have broken  card-sc-hsm.c too. 

https://github.com/OpenSC/OpenSC/commit/e5707b545e5a2dc33b0ca52a8bf63f36f71b3d85 was adding "Add support for PSS padding to RSA signatures", so it is not clear how it would break any EC code that does not use padding.  

Your fix may work, but it looks like it there is something more fundamental wrong that should be fixed. 
OpenSC has two flags  SC_ALGORITHM_ECDSA_RAW and  SC_ALGORITHM_ECDH_CDH_RAW.  There may be keys that which can only do one but not both. So there should be a reason for two different flags. And it would make more sense if a derive was being done, it would use SC_ALGORITHM_ECDH_CDH_RAW.  

git blame can be used to look at which commits introduced/modified the two flags. I know ECDSA was added to OpenSC before ECDH.  Maybe the author of https://github.com/OpenSC/OpenSC/commit/e5707b545e5a2dc33b0ca52a8bf63f36f71b3d85 can comment.

**Comment by frankmorgner** [comment]
... that would be @NWilson ...

**Comment by dengert** [comment]
This is just a though, I have not looked closely at this. I would question the changes to the definitions of `SC_ALGORITHM_RSA_PAD_NONE` and `SC_ALGORITHM_RSA_HASH_NONE` from 0 (i.e an absent of any PAD or HASH) to 1 (treating a NONE as its own type of PAD  or HASH):

From: `#define SC_ALGORITHM_RSA_PADS		0x0000001E`

To:   `#define SC_ALGORITHM_RSA_PADS		0x0000001F`

From: `#define SC_ALGORITHM_RSA_PAD_NONE	0x00000000`
To:     `#define SC_ALGORITHM_RSA_PAD_NONE	0x00000001`

#define SC_ALGORITHM_RSA_RAW		0x00000001


Addition of: #define SC_ALGORITHM_RSA_HASH_NONE	0x00000100

From: #define SC_ALGORITHM_RSA_HASHES		0x0001FE00
To:   #define SC_ALGORITHM_RSA_HASHES		0x0001FF00

The ECDSA code later has this: 

#define SC_ALGORITHM_ECDSA_HASH_NONE		SC_ALGORITHM_RSA_HASH_NONE

This could be the problem, as the ECDSA code that tests `SC_ALGORITHM_ECDSA_HASH_NONE` many need similar changes, or the RSA code may need to be reverted.

**Comment by Jakuje** [comment]
Well, the above referenced commit was actually brought in by me after the initial draft by @NWilson years ago. Unfortunately, I do not have any myeid card to try this, but I can try with the sc-hsm if I will find the card and be able to reproduce the issue.

**Comment by popovec** [comment]
In current code `priv->sec_env->algorithm_flags` is set to 0 (in `sc_get_encoding_flags()`).  Any of card driver may need info about ECC operation (ECDSA/ECDH). A this moment this information is not available. (Alternative -  it is not necessary to manage this flag, operation can be deduced by card driver function - if `sign` or `derive` function is called).

The question is, where the `SC_ALGORITHM_ECDSA_RAW` / `SC_ALGORITHM_ECDH_CDH_RAW` flag should be set. In previous code, this flag was set in `sc_get_encoding_flags()`.  IMHO better way is to  set this flag in function, that really knows what type of operation is requested from card driver  - in this case in `pkcs15_prkey_derive()` - as suggested above in proposed patch. Flag `SC_ALGORITHM_ECDSA_RAW`should be set in  `pkcs15_prkey_sign()` (in cases   CKM_ECDSA and CKM_ECDSA_SHA1). 

The MyEID card driver should be updated too, `priv->sec_env->algorithm_flags & SC_ALGORITHM_ECDSA_RAW` should be rewritten to ` priv->sec_env->algorithm_flags & SC_ALGORITHM_ECDH_CDH_RAW` (as suggested  by @dengert ).

**Comment by Jakuje** [comment]
Tested with sc-hsm and it looks like I have the same problem with master, while it works fine with 0.19.0.

I will investigate what needs to be done to fix this and I will try to integrate some of the derive testing to the testsuites to avoid these issues in future.

**Comment by Jakuje** [comment]
From what I understand the code, the detection was fundamentally broken already previously since the problematic code in `sc_get_encoding_flags()` basically in the past pulled all the raw algorithms the card supports and passed them as sFlags  (see the `SC_ALGORITHM_ECDSA_RAW` capability appearing out of nowhere while the iFlags in snippets in description is empty).

The change that broke this was

**Comment by dengert** [comment]
You said above: "Any of card driver may need info about ECC operation (ECDSA/ECDH)" That appears to be correct in the call to  `piv_set_security_env`, the `env->operation` is `SC_SEC_OPERATION_DERIVE` and  
`env->algorithm  `SC_ALGORITHM_EC` But that maybe enough for the PIV that can only do one type of derive. 

`pkcs11-tool` with `--derive -m ECDH1-COFACTOR-DERIVE`  works with current master. And the sflags look OK as there is no padding, Here is part of an opensc-debug.log:
[piv-derive.txt](https://github.com/OpenSC/OpenSC/files/3189527/piv-derive.txt)

**Comment by Jakuje** [comment]
@dengert If I read your last comment correctly, you say that the set_security_environment() of the sc-hsm and myeid is wrong and it should first look to the operation and than to the algorithm flags?

**Comment by dengert** [comment]
Maybe. 

OpenSC pkcs11.h defines these "DERIVE"  mechanisums:
```
#define CKM_DH_PKCS_DERIVE		(0x21UL)
#define CKM_X9_42_DH_DERIVE		(0x31UL)
#define CKM_X9_42_DH_HYBRID_DERIVE	(0x32UL)
#define CKM_X9_42_MQV_DERIVE		(0x33UL)
#define CKM_SSL3_MASTER_KEY_DERIVE	(0x371UL)
#define CKM_SSL3_KEY_AND_MAC_DERIVE	(0x372UL)
#define CKM_SSL3_MASTER_KEY_DERIVE_DH	(0x373UL)
#define CKM_TLS_MASTER_KEY_DERIVE	(0x375UL)
#define CKM_TLS_KEY_AND_MAC_DERIVE	(0x376UL)
#define CKM_TLS_MASTER_KEY_DERIVE_DH	(0x377UL)
#define CKM_KEA_KEY_DERIVE		(0x1011UL)
#define CKM_ECDH1_DERIVE		(0x1050UL)
#define CKM_ECDH1_COFACTOR_DERIVE	(0x1051UL)
#define CKM_ECMQV_DERIVE		(0x1052UL)
#define CKM_GOSTR3410_DERIVE    (0x1204UL)
#define CKM_GOSTR3410_12_DERIVE	(CK_VENDOR_PKCS11_RU_TEAM_TK26 | 0x007)
```

So for ECDH there are 3. 

`pkcs11-tool.c` can handle all 3.
  
`framework-pkcs15.c` does not support `CKM_ECMQV_DERIVE` or is parameters. 
` CKM_ECDH1_COFACTOR_DERIVE` and `CKM_ECDH1_COFACTOR_DERIVE` use the same parameters i.e. `CK_ECDH1_DERIVE_PARAMS` containing the pubkey of the other party which is passed to `sc_pkcs15_derive` as the `pSeedData`.

The choice of mechanism is not passed  to `sc_pkcs15_derive`. So this may be a problem. 

 The original ECDH code was written for the PIV where `CKM_ECDH1_DERIVE` and  `CKM_ECDH1_COFACTOR_DERIVE` produce the same results and these was no need to pass in which of the 3 mechanisms was requested. 

`framework-pkcs15.c` could pass the mechanism to   `sc_pkcs15_derive` maybe as  `sc_algorithm_info_t` which would also contain the  `pSeedData` or the full eparameters.

**Comment by dengert** [comment]
One other note. `card-myeid.c` contains what looks like a bug:
```
1292         if (priv->sec_env && priv->sec_env->algorithm == SC_ALGORITHM_EC
1293                 && priv->sec_env->operation == SC_SEC_OPERATION_DERIVE
1294                 && priv->sec_env->algorithm_flags & SC_ALGORITHM_ECDSA_RAW)
```
Should the last line be:
```
1294                 && priv->sec_env->algorithm_flags & SC_ALGORITHM_ECDH_CDH_RAW)
```
Do we need to define all three ECDH mechanisms here?

**Comment by mouse07410** [comment]
@dengert I'd say yes and yes.

**Comment by popovec** [comment]
`priv->sec_env->algorithm_flags` is in current code always 0,  changing this test from `SC_ALGORITHM_ECDSA_RAW` to any different value does not fix this issue. Because `CKM_ECDH1_COFACTOR_DERIVE` and `CKM_ECDH1_DERIVE` is  same operation (src/tools/pkcs11-tool.c).. we can remove `priv->sec_env->algorithm_flags & SC_ALGORITHM_ECDSA_RAW` test from MyEID driver. sc-hsm driver is to be fixed by similar change.

**Comment by mouse07410** [comment]
> `CKM_ECDH1_COFACTOR_DERIVE` and `CKM_ECDH1_DERIVE` is same operation (`src/tools/pkcs11-tool.c`)

It may be the same *now*. It may change in the future.

> is in **current code** always 0

What if the next release changes this? I say this test should be fixed properly.

**Comment by dengert** [comment]
`CKM_ECDH1_COFACTOR_DERIVE` and `CKM_ECDH1_DERIVE` are not same operation in `pkcs11-tool.c`:
```
3810         switch(opt_mechanism) {
3811         case CKM_ECDH1_COFACTOR_DERIVE:
3812         case CKM_ECDH1_DERIVE:
3813                 derived_key= derive_ec_key(session, key, opt_mechanism);
```
so the `opt_mechanism` is passed to derive_ec_key as `mech_mech` then:
```
3678 derive_ec_key(CK_SESSION_HANDLE session, CK_OBJECT_HANDLE key, CK_MECHANISM_TYPE mech_mech)
...
3715         mech.mechanism = mech_mech;
...
3779         rv = p11->C_DeriveKey(session, &mech, key, newkey_template, n_attrs, &newkey);
```

So pkcs11-tool.c supports the 2 mechanisms but not `CKM_ECMQV_DERIVE`.

But OpenSC pkcs11 does not do anything with  the passed mechanism. 

I don't think we have any cards that support `CKM_ECMQV_DERIVE`. So for now we can ignore it. 

NIST 800-73-4 says in part 2:
"A.5.2.1 ECDH with the PIV KMK
The PIV Card supports ECDH key agreement by performing the elliptic curve cryptography cofactor Diffie-Hellman (ECC CDH) primitive [SP800-56A, Section 5.7.1.2] using its ECC KMK private key and an ECC public key that is provided as input to the GENERAL AUTHENTICATE command. All other procedures required to complete key agreement are performed by the cardholder’s client application and its associated cryptographic module."

So it only supports ECC CDH i.e. `CKM_ECDH1_COFACTOR_DERIVE`. 

See next comment on what we could do for other cards...

**Comment by dengert** [comment]
@popovec As you point  "we can remove priv->sec_env->algorithm_flags & SC_ALGORITHM_ECDSA_RAW"
as the flag is never set. This only works if the myeid card only supports 1 type of ECDH, much like teh PIV card does. 

It the `sc-hsm` used to work we could do something similar. 


**But If we wanted to fix this for other cards or for future cards** we could look at setting the `priv->sec_env->algorithm_flags` and defining flags. This would then inform the card driver what the user requested. OpenSC defines `SC_ALGORITHM_ECDH_CDH_RAW` which looks like it this matches `CKM_ECDH1_COFACTOR_DERIVE` but does not define a `SC_ALGORITHM_ECDH_RAW` which would match `CKM_ECDH1_DERIVE`.

All cards supporting ECDH should then be changed to check  for a supported `SC_ALGORITHM_ECDH_*`

What do you think?

**Comment by popovec** [comment]
@dengert ok, we need to introduce  at minimum two flags`SC_ALGORITHM_ECDH_CDH_RAW` and `SC_ALGORITHM_ECDH_RAW`.
According to MyEID doc (2.1.4), MyEID need check here for `SC_ALGORITHM_ECDH_RAW` not `SC_ALGORITHM_ECDH_CDH_RAW` (but someone from Aventra must confirm this) . Lot of card drivers are affected by this change: card-sc-hsm.c, card-piv.c, card-openpgp.c, card-myeid.c, card-mcrd.c, card-esteid2018.c, card-coolkey.c. We need to check lot of card docs, what does really the card ? ECDH key derivation operation  or ECDH with COFACTOR  key derivation. (Cofactor operation is more secure, if we do not known, what operation the card does, there is better to check `SC_ALGORITHM_ECDH_RAW` in card driver.)

**Comment by dengert** [comment]
Doing a Google search for "ECDH COFACTOR" shows that the cofactor is one of the curve parameters "h". 

[pkcs11-curr-v2.40-cs01](http://docs.oasis-open.org/pkcs11/pkcs11-curr/v2.40/cs01/pkcs11-curr-v2.40-cs01.html) says:
"2.3.9 Elliptic curve Diffie-Hellman key derivation
The elliptic curve Diffie-Hellman (ECDH) key derivation mechanism, denoted CKM_ECDH1_DERIVE, is a mechanism for key derivation based on the **Diffie-Hellman version** of the elliptic curve key agreement scheme, **as defined in ANSI X9.63**, where each party contributes one key pair all using the same EC domain parameters."

"2.3.10 Elliptic curve Diffie-Hellman with cofactor key derivation
The elliptic curve Diffie-Hellman (ECDH) with cofactor key derivation mechanism, denoted CKM_ECDH1_COFACTOR_DERIVE, is a mechanism for key derivation based on the **cofactor Diffie-Hellman version** of the elliptic curve key agreement scheme, **as defined in ANSI X9.63**, where each party contributes one key pair all using the same EC domain parameters.  Cofactor multiplication is computationally efficient and helps to prevent security problems like small group attacks."

Both referring to "ANSI X9.63"  But I don't have a copy of ASNI X9.63!

So I am speculating that the only difference is to multiple by "h". So for curves with h=1 the results are the same. This may not be true, so someone with a copy of ASNI X9.63 needs to look at this.

Looking at OpenSSL source appears to confirm this, but I have not tries any tests. 

SInce "Cofactor multiplication is computationally efficient and helps to prevent security problems" I would also speculate that all smart cards would be using Cofactor multiplication.

The default for card drivers appears to do what they do now. Some don't check (card-piv.c as an example) others are referring to  "SC_ALGORITHM_ECDH_CDH_RAW" that implies they support Cofactor.

But if we differentiate between  "CKM_ECDH1_DERIVE" and "CKM_ECDH1_COFACTOR_DERIVE" there could be issues with existing code. The card driver would have to make the call. And for curves with h=1 if I am speculating correctly there is no difference. 

[NIST.SP.800-56Ar3](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-56Ar3.pdf) Section 5.7 says: 
"2. The ECC CDH primitive (called the Modified Diffie-Hellman primitive in ANS X9.63; see Section 5.7.1.2 below): This primitive shall be used by the Full Unified Model, Ephemeral Unified Model, One-Pass Unified Model, One-Pass DiffieHellman and Static Unified Model schemes, which are based on elliptic curve cryptography and the Diffie-Hellman algorithm."

I agree  with your statment: "is better to check SC_ALGORITHM_ECDH_RAW in card driver."

If you want I could work up a mod to modify `framework-pkcs15.c` and `sc_pkcs15_derive`.

**Comment by dengert** [comment]
A quick look at the OpenSC code presented as a spreed sheet below of supported curves and cards.

- "x" indicates the card driver or pkcs15-card file lists named curves and lengths.
- "?" indicates it supports some curve of the specific length, I assume these are for the most common  curves. 
- All curves have a cofactor of 1.
- No driver distinguishes between different types of ECDH. i.e. only has one ADPU command for ECDH.
- OpenPGP says it supports 512 as  ansiX9p512r1, but the OID is for [1.3.132.0.35](http://oid-info.com/get/1.3.132.0.35) So curve name in the comment is wrong and length is wrong.  RFC 6637 says OpenPGP will support OID "1.3.132.0.35" "NIST curve P-521" 
- There is no common namedCurve across all cards, Most support secp256r1 or secp384r1. If sc-hsm supported secp384r1 there would be a common common curve. This may only be software, as it looks like sc-hsm looks up a curve and sends the parameters to card. Maybe it could also support secp521r1 too? 

Comments? 




Curve | h | myeid | isoApplet | sc-hsm | openpgp | PIV | coolkey | mcrd | esteid2018
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
secp192r1 | 1 | x | x | x |   |   |   |   |  
secp256r1 | 1 | x | x | x | x | x | ? |   |  
secp384r1 | 1 | x | x |   | x | x | ? | ? | ?
ansiX9p512r1 | ? |   |   |   | misprint |   |   |   |  
secp521r1 | 1 | x |   |   | x OID |   | ? |   |  
secp224r1 | 1 |   | x |   |   |   |   |   |  
secp192k1 | 1 |   | x |   |   |   |   |   |  
secp256k1 | 1 |   | x |   |   |   |   |   |  
brainpoolP192r1 | 1 |   | x | x |   |   |   |   |  
brainpoolP224r1 | 1 |   | x | x |   |   |   |   |  
brainpoolP256r1 | 1 |   | x | x | x |   |   |   |  
brainpoolP320r1 | 1 |   | x | x |   |   |   |   |  
brainpoolP384r1 | 1 |   |   | x | x |   |   |   |  
brainpoolP512r1 | 1 |   |   | x |   |   |   |   |

**Comment by popovec** [comment]
If different curve names corresponds to same OID, we can assume, this is same curve (calculation is running with same curve parameters)

nistp192/prime192v1/secp192r1  = OID 1.2.840.10045.3.1.1
nistp256/prime256v1/secp256r1 = OID 1.2.840.10045.3.1.7 

`ansiX9p512r1` I assume typo error here, 
ansip521r1/secp521r1 =  OID 1.3.132.0.35
 
To cofactor derive operation....  here is some mathematical explanation about this:

https://crypto.stackexchange.com/questions/18222/difference-between-ecdh-with-cofactor-key-and-ecdh-without-cofactor-key

I am not sure, if I can interpret this correctly, Alice generate a secret key _a_ on card, then _aG_ is corresponding public key.  This public key is then used to generate shared secret with Bob, Bob uses private key _b_ to compute _b(aG)_.  This normal ECDH operation. If cofactor ECDH is requested, this make sense only in case if curve parameter _h!=1_ . Thanks to @dengert , here is table - all cards supported by opensc uses curves with _h=1_. ECDH and ECDH-COFACTOR is for now the same operation. 

If above interpretation is without errors, it doesn't matter if flag used for ECDH is named `SC_ALGORITHM_ECDH_RAW` or `SC_ALGORITHM_ECDH_CDH_RAW` (for now)... There is problem, in which part of the code we need set this flag. Is `sc_get_encoding_flags()` the correct place ? Or all flags needed for ECDH/ECDSA  should be generated in  _src/pkcs11/framework-pkcs15.c_, functions `pkcs15_prkey_sign()` and `pkcs15_prkey_derive()`.

Or, we can remove this flag, because `env->operation == SC_SEC_OPERATION_DERIVE` is for now enough to check type of requested operation. In future, maybe it is enough to define `SC_SEC_OPERATION_COFACTOR_DERIVE`.

Card APDU - there is same APDU for ECDH and ECDH-COFACTOR (general auth, 00h 86h 00h 00h XXh )? Is there place in data field - for example in Dynamic Authentication Template  to pass operation type to card ?  Or card driver may use _set security environment_  operation to pass type of operation to card (in some tag of Auth template). Can somebody check this in ISO7816-X ? I can't find this in iso7816-4, here only tag 0x80 - Algorithm reference with value  X4h  is used for  ECDSA/ECDH. Upper part of value is for EC always 0 - there is no hash algo.

IMHO, It is better to fix this error in the easiest way. If some of card driver in future will need more parameters, this can be handled by separate patch.

**Comment by dengert** [comment]
Having been looking at this problem all day.   One issue is `format_senv` is called in 5 places to setup the `senv_out` and `alg_info_out`. Only after `format_senv` completes, the calling routine sets  `senv.operation = SC_SEC_OPERATION_{SIGN|DECIPHER|DERIVE|UNWRAP|WRAP}`

`format_senv` is making the assumption that for each type of key these is only one type of algorithm. (It does search on the `prkey->field_length`.)   Works for  `SC_ALGORITHM_RSA` and `SC_PKCS15_TYPE_SKEY_GENERIC`,  may not work for  `SC_ALGORITHM_GOSTR3410` and does not for `SC_ALGORITHM_EC`.

EC KEYS could have 4 types of possible algorithms: `ECDSA: SC_ALGORITHM_EC`,  `ECDH_DERIVE:(Not defined yet)`, `ECDH_COFACTOR_DERIVE: SC_ALGORITHM_ECDH_CDH_RAW` and `EC_MQV_DERIVE: (Not supported yet)`.  

So I am looking at passing to`format_senv`  the `SC_SEC_OPERATION_*` and for SC_SEC_OPERATION_DERIVE the requested type of derive:   SC_ALGORITHM_ECDH_CDH_RAW or  SC_ALGORITHM_ECDH_RAW. 

`format_senv` will then  rather then using `sc_card_find_ec_alg` which assumes `SC_ALGORITHM_EC` `format_senv` will use `sc_card_find_alg` which can also search for `curve_name` too. (May require  a change to _sc_card_add_ec_alg which sets `info.algorithm = SC_ALGORITHM_EC;`) 

@NWilson  (or anyone else)  can you verify  if we a problem with the following: 
```
#define SC_ALGORITHM_ECDH_CDH_RAW	0x00200000       e2f0e367b (Leif Erik Wagner 2018-06-05
#define SC_ALGORITHM_ECDSA_RAW		0x00100000       e5707b545 (Nicholas Wilson 2015-08-25
```
conflict with:
```
/* This defines the hashes to be used with MGF1 in PSS padding */
#define SC_ALGORITHM_MGF1_SHA1          0x00100000      e5707b545 (Nicholas Wilson  2015-08-25
#define SC_ALGORITHM_MGF1_SHA256        0x00200000   e5707b545 (Nicholas Wilson  2015-08-25
``


We are running out of bits in the  `SC_ALGORITHM_*` space. We may only have these bits left 0x3e000000 
I want to add 1, move the 2 that conflict and GOST may need one. 

Another reason to not have added the `SC_ALGORITHM_RSA_HASH_NONE` and `SC_ALGORITHM_RSA_PAD_NONE` as separate bits rather then defined as no other has or pad bits set.

**Comment by Jakuje** [comment]
I am following on the backlog in this issue. Indeed, theoretically the above flags conflict, which is probably wrong, but I either missed that back then or did not consider that big enough issue (since these flags are somehow separate for different key types and operations stored in the security_environment structure).

Thank you for the analysis and references. It somehow matches what I figured out during last days (since I did not know a lot about ECDH either), but I plan to spend some more time on this during coming days checking how does it work in the testsuite.

**Comment by Jakuje** [comment]
In #1677 I tried to fix this issue and implemented also a simple derive test that seems to work for me now against both softhsm and sc-hsm card I have around.

Now, I am going to have a look into the software implementation of myeid as proposed by @popovec since it is the only MyEID card I can try, but testing of my branch with real card would be appreciated too.  I will try other cards later too.

**Comment by dengert** [comment]
#1667 looks good. 

Sitll have 2 misprints in OpenPGP of curce name and size of a 521 bit curve. 
```
163         {{{1, 3, 132, 0, 35, -1}}, 512,
164                 {{0x2b, 0x81, 0x04, 0x00, 0x23, -1}}}, /* ansiX9p512r1 */
```
should be:
```
163         {{{1, 3, 132, 0, 35, -1}},  521,
164                 {{0x2b, 0x81, 0x04, 0x00, 0x23, -1}}}, /*  ansip521r1  and secp521r1 */
```

**Comment by Jakuje** [comment]
@dengert yes, the opengpg card is indeed a typo. This number is very prone to mistype. I added a commit to #1677 to fix this also. My OpenSSL does not have brainpool curves so I can not check anything else with them and I had also some issues with `secp192r1` while testing with the OsEID (MyEID simulator).

I tested also the derive with NIST PIV cards and that seems to work fine. For the Coolkey cards, the support is actually there, but I do not have any card having the EC keys so I can not test that. The same for the isoApplet, but that one should be possible to make working in the Travis CI.

While integrating the OsEID simulator to the travis train, I was thinking whether it would make sense to separate the runs for different cards so it does not take that long and is more clear what got broken from the results.

**Comment by dengert** [comment]
Thanks for adding the OpenPGP fix. 

One other item. A card may support ECC keys with ECDSA but not with ECDH. I believe some of the earlier Yubico cards may have had this restriction.  The current "sc_pkcs15" code treats the SC_ALGORTHM_EC as if it was one algorithm and uses the "operation" to distinguish between ECDSA, ECDH  and ECDH_CDH. The "pkcs11" register mechanisms sees SC_ALGORTHM_EC and returns PKCS11 mechanisms. 

`pkcs11-tool -M`  includes:
```
 ECDSA, keySize={256,384}, hw, sign, other flags=0x1800000
 ECDH1-COFACTOR-DERIVE, keySize={256,384}, hw, derive, other flags=0x1800000
 ECDH1-DERIVE, keySize={256,384}, hw, derive, other flags=0x1800000
```
We should consider replacing `SC_ALGORTHM_EC` with  3 separate algorithms, and have the drivers regester each if they support them. This will then make it easier to add ECDH_MQV_DERIVE at a later time too.

PKCS\#15 v1.1  says in  "TokenInfo.supportedAlgorithms"  "The intent of this optional field is to indicate cryptographic algorithms, associated parameters, operations and algorithm input formats. 
...
Values for the algorithm field shall be chosen from, and interpreted as, mechanism numbers in PKCS\#11
... 
Values of the supportedOperations field (compute-checksum, compute-signature, verify-checksum, verify-signature, encipher, decipher, hash and derive-key) identifies operations the card can perform with a particular algorithm."

PKCS\#15 was last updated in 2000 (I have not found newer versions) and it referres to the "operations". But PKCS\#11 is still being updated and treats ECDSA, ECDH and ECDH_CDH as separate mechanisms. 

We also have many new PKCS\#11 mechanisms that were not around in 2000, such as GOST. I don't use GOST but it looks like it could stand some updating similiar to ECC.

**Comment by Jakuje** [comment]
According to the wiki [1], the PKCS#15 specification was updated with ISO/IEC 7816-15, which was last updated in 2016, though not sure what is in.

From what I know, the GOST cards are subject to Russia export control (similarly as the CAC cards in US) so they can not be used anywhere else in the world so we as developers can hardly do anything to improve their support in our software.

I would consider the algorithm separation as a separate issue, but indeed, I can have a look what can be done about that later. In PKCS#11, there are always the private key attributes, that define whether this mechanism can be used with key (`CKA_DERIVE` in our case), which are strictly checked in our PKCS#11 object before even trying to do the derive:

https://github.com/OpenSC/OpenSC/blob/master/src/pkcs11/pkcs11-object.c#L1220-1224

Indeed, having a token-level/driver overview about what algorithms should be announced in the mechanisms list would be good.

I will try to refactor the travis ci job so the emulators do not conflict and pass at least.


[1] https://en.wikipedia.org/wiki/PKCS

**Comment by popovec** [comment]
fixed by #1677
